File: ecs-hosted/go/application/service/consoleService.go
package service

import (
	"encoding/json"
	"fmt"

	"github.com/pulumi/pulumi-aws/sdk/v4/go/aws/ec2"
	"github.com/pulumi/pulumi-aws/sdk/v4/go/aws/lb"
	"github.com/pulumi/pulumi-self-hosted-installers/ecs-hosted/application/log"
	"github.com/pulumi/pulumi-self-hosted-installers/ecs-hosted/application/network"
	"github.com/pulumi/pulumi-self-hosted-installers/ecs-hosted/application/utils"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

const consolePort = 3000
const consoleContainerName = "pulumi-service"

func NewConsoleContainerService(ctx *pulumi.Context, name string, args *ConsoleContainerServiceArgs, opts ...pulumi.ResourceOption) (*ConsoleContainerService, error) {
	var resource ConsoleContainerService

	err := ctx.RegisterComponentResource("pulumi:consoleService", name, &resource, opts...)
	if err != nil {
		return nil, err
	}

	// create our parented options
	options := append(opts, pulumi.Parent(&resource))

	listenerConditions := &lb.ListenerRuleConditionArray{
		lb.ListenerRuleConditionArgs{
			HostHeader: lb.ListenerRuleConditionHostHeaderArgs{
				Values: pulumi.StringArray{pulumi.String(args.ConsoleUrl)},
			},
		},
		lb.ListenerRuleConditionArgs{
			PathPattern: lb.ListenerRuleConditionPathPatternArgs{
				Values: pulumi.StringArray{pulumi.String("/*")},
			},
		},
	}

	taskArgs, err := newConsoleTaskArgs(ctx, args)
	if err != nil {
		return nil, err
	}

	tgName := fmt.Sprintf("%s-tg", name)
	tgOptions := append(options, pulumi.DeleteBeforeReplace(true))
	tg, err := lb.NewTargetGroup(ctx, tgName, &lb.TargetGroupArgs{
		VpcId:      args.VpcId,
		Protocol:   pulumi.String("HTTP"),
		Port:       pulumi.Int(consolePort),
		TargetType: pulumi.String("ip"),
		HealthCheck: &lb.TargetGroupHealthCheckArgs{
			Interval:           pulumi.Int(10), //seconds
			Path:               pulumi.String("/"),
			Port:               pulumi.String(fmt.Sprintf("%d", consolePort)),
			Protocol:           pulumi.String("HTTP"),
			Matcher:            pulumi.String("200-299"),
			Timeout:            pulumi.Int(5), //seconds
			HealthyThreshold:   pulumi.Int(5),
			UnhealthyThreshold: pulumi.Int(2),
		},
	}, tgOptions...)

	if err != nil {
		return nil, err
	}

	httpsListener, err := args.TrafficManager.Public.CreateListenerRule(ctx, fmt.Sprintf("%s-https", name), true, tg.Arn, listenerConditions, options...)
	if err != nil {
		return nil, err
	}

	httpListener, err := args.TrafficManager.Public.CreateListenerRule(ctx, fmt.Sprintf("%s-http", name), false, tg.Arn, listenerConditions, options...)
	if err != nil {
		return nil, err
	}

	serviceOptions := append(options, pulumi.DependsOn([]pulumi.Resource{httpsListener, httpListener}))
	resource.ContainerService, err = NewContainerService(ctx, name, &ContainerServiceArgs{
		ContainerBaseArgs:          args.ContainerBaseArgs,
		PulumiLoadBalancer:         args.TrafficManager.Public,
		PulumiInternalLoadBalancer: args.TrafficManager.Internal,
		TargetPort:                 consolePort,
		TaskDefinitionArgs:         taskArgs,
		TargetGroups:               []*lb.TargetGroup{tg},
	}, serviceOptions...)

	if err != nil {
		return nil, err
	}

	sgOptions := append(options, pulumi.DeleteBeforeReplace(true))
	_, err = ec2.NewSecurityGroupRule(ctx, fmt.Sprintf("%s-lb-to-ecs-rule", name), &ec2.SecurityGroupRuleArgs{
		Type:                  pulumi.String("egress"),
		SecurityGroupId:       args.TrafficManager.Public.SecurityGroup.ID(),
		SourceSecurityGroupId: resource.ContainerService.SecurityGroup.ID(),
		FromPort:              pulumi.Int(consolePort),
		ToPort:                pulumi.Int(consolePort),
		Protocol:              pulumi.String("TCP"),
		Description:           pulumi.String("Allow access from UI LB to ecs service"),
	}, sgOptions...)

	if err != nil {
		return nil, err
	}

	if err != nil {
		return nil, err
	}

	return &resource, nil
}

func newConsoleTaskArgs(ctx *pulumi.Context, args *ConsoleContainerServiceArgs) (*TaskDefinitionArgs, error) {
	// set out defaults for the console container task(s)
	taskMemory := 512
	if args.TaskMemory > 0 {
		taskMemory = args.TaskMemory
	}

	taskCpu := 256
	if args.TaskCpu > 0 {
		taskCpu = args.TaskCpu
	}

	containerMemoryRes := 128
	if args.ContainerMemoryReservation > 0 {
		containerMemoryRes = args.ContainerMemoryReservation
	}

	containerCpu := taskCpu
	if args.ContainerCpu > 0 {
		containerCpu = args.ContainerCpu
	}

	numberDesiredTasks := 3
	if args.DesiredNumberTasks > 0 {
		numberDesiredTasks = args.DesiredNumberTasks
	}

	ecrAccountId := args.AccountId
	if args.EcrRepoAccountId != "" {
		ecrAccountId = args.EcrRepoAccountId
	}

	imageName := fmt.Sprintf("pulumi/console:%s", args.ImageTag)
	fullQualifiedImage := utils.NewEcrImageTag(ecrAccountId, args.Region, imageName, args.ImagePrefix)

	// resolve all needed outputs to construct our container definition in JSON
	conatinerDefinitions, _ := pulumi.All(
		args.TrafficManager.Public.LoadBalancer.DnsName,
		args.LogDriver).ApplyT(func(applyArgs []interface{}) (string, error) {

		dnsName := applyArgs[0].(string)
		logDriver := applyArgs[1].(log.LogDriver)

		containerJson, err := json.Marshal([]interface{}{
			map[string]interface{}{
				"cpu":               containerCpu,
				"environment":       newConsoleEnvironmentVariables(args, dnsName),
				"image":             fullQualifiedImage,
				"logConfiguration":  logDriver.GetConfiguration(),
				"memoryReservation": containerMemoryRes,
				"name":              consoleContainerName,
				"portMappings": []map[string]interface{}{
					{
						"containerPort": consolePort,
					},
				},
			},
		})

		if err != nil {
			return "", err
		}

		return string(containerJson), nil
	}).(pulumi.StringOutput)

	return &TaskDefinitionArgs{
		ContainerDefinitions: conatinerDefinitions,
		NumberDesiredTasks:   numberDesiredTasks,
		Cpu:                  taskCpu,
		Memory:               taskMemory,
		ContainerName:        consoleContainerName,
		ContainerPort:        consolePort,
	}, nil
}

func newConsoleEnvironmentVariables(args *ConsoleContainerServiceArgs, lbDnsName string) []map[string]interface{} {

	env := []map[string]interface{}{
		CreateEnvVar("AWS_REGION", args.Region),
		CreateEnvVar("LOGIN_RECAPTCHA_SITE_KEY", args.RecaptchaSiteKey),
		CreateEnvVar("PULUMI_API", fmt.Sprintf("https://%s", args.ApiUrl)),
		CreateEnvVar("PULUMI_CONSOLE_DOMAIN", args.ConsoleUrl),
		CreateEnvVar("PULUMI_HOMEPAGE_DOMAIN", args.ConsoleUrl),
		CreateEnvVar("PULUMI_ROOT_DOMAIN", args.RootDomain),
		CreateEnvVar("RECAPTCHA_SITE_KEY", args.RecaptchaSiteKey),
	}

	// this URL should correspond to a route53 A record which aliases the internal private NLB
	// this is only needed when egress on the ECS service SG is locked down to VPC CIDR
	// ALB's don't have static IP addresses, hence, the NLB
	if args.EnablePrivateLoadBalancerAndLimitEgress {
		env = append(env, CreateEnvVar("PULUMI_API_INTERNAL_ENDPOINT", fmt.Sprintf("https://%s", args.ApiInternalUrl)))
	}

	if args.HideEmailLogin {
		env = append(env, CreateEnvVar("PULUMI_HIDE_EMAIL_LOGIN", "true"))
	}

	if args.HideEmailLogin {
		env = append(env, CreateEnvVar("PULUMI_HIDE_EMAIL_SIGNUP", "true"))
	}

	if args.SamlSsoEnabled {
		env = append(env, CreateEnvVar("SAML_SSO_ENABLED", "true"))
	}

	return env
}

type ConsoleContainerServiceArgs struct {
	ContainerBaseArgs

	ApiUrl                     string
	ApiInternalUrl             string
	ConsoleUrl                 string
	ContainerCpu               int
	ContainerMemoryReservation int
	DesiredNumberTasks         int
	EcrRepoAccountId           string
	HideEmailSignup            bool
	HideEmailLogin             bool
	ImageTag                   string
	ImagePrefix                string
	LogDriver                  log.LogDriver
	RecaptchaSiteKey           string
	RootDomain                 string
	SamlSsoEnabled             bool
	TaskMemory                 int
	TaskCpu                    int
	TrafficManager             *network.TrafficManager
	WhiteListCidrBlocks        []string
}

type ConsoleContainerService struct {
	pulumi.ResourceState

	ContainerService *ContainerService
}

File: ecs-hosted/ts/application/services/consoleService.ts
import * as pulumi from "@pulumi/pulumi";
import { ec2 } from "@pulumi/aws/";
import { input } from "@pulumi/aws/types";

import { ContainerService } from "./containerService";
import { TaskDefinitionArgs, ConsoleServiceArgs } from "./types";
import { buildECRImageTag } from "../utils";
import { LogType, LogDriver } from "../logs/types";
import { LogFactory } from "../logs/logFactory";

const namespace = "pulumi:consoleService";
const consolePort = 3000;
const consoleContainerName = "pulumi-console";

/*
Represents the base of the Pulumi Console.
Comprised of an ECS Cluster/Service and all required pieces (tasks, containers, etc)
Tasks DO NOT need access to Aurora DB and should NOT access S3 buckets
*/
export class ConsoleService extends pulumi.ComponentResource {

    private readonly consoleService: ContainerService;
    private readonly baseArgs: ConsoleServiceArgs;
    private readonly options: pulumi.ComponentResourceOptions;
    private readonly consoleImage: string;

    constructor(name: string, args: ConsoleServiceArgs, opts?: pulumi.ComponentResourceOptions) {
        super(namespace, name, args, opts);

        this.baseArgs = args;
        this.options = pulumi.mergeOptions(opts, { parent: this });
        this.consoleImage = `pulumi/console:${this.baseArgs.imageTag}`;

        const { trafficManager } = args;

        // specify the health check to be utilized with Pulumi Console
        const healthCheck: input.lb.TargetGroupHealthCheck = {
            interval: 10,  // seconds
            path: "/",
            protocol: "HTTP",
            matcher: "200-299",
            timeout: 5,  // seconds
            healthyThreshold: 5,
            unhealthyThreshold: 2,
        };

        // listener conditions allow ALB to route appropriate traffic to Pulumi Console
        // traffic should match app.pulumi.whatever as well as pulumi.whatever
        const listenerConditions = [
            {
                hostHeader: {
                    values: [
                        args.dns.consoleUrl, 
                        trafficManager.console.loadBalancer.dnsName, 
                        this.baseArgs.dns.rootDomain
                    ]
                }
            },
            {
                pathPattern: {
                    values: ["/*"]
                }
            }
        ];

        // Pulumi Console should never communicate directly with Pulumi API. Traffic should always route through ALB (public)
        const taskArgs = this.constructTaskArgs();
        this.consoleService = new ContainerService(`${name}`, {
            region: args.region,
            accountId: args.accountId,
            cluster: args.cluster,
            healthCheck: healthCheck,
            listenerConditions: listenerConditions,
            listenerPriority: 1,
            privateSubnetIds: args.privateSubnetIds,
            pulumiLoadBalancer: trafficManager.console,
            targetPort: consolePort,
            taskDefinitionArgs: taskArgs,
            vpcId: args.vpcId,
        }, this.options);

        new ec2.SecurityGroupRule(`${name}-lb-to-console-rule`, {
            type: "egress",
            securityGroupId: trafficManager.console.securityGroup.id,
            sourceSecurityGroupId: this.consoleService.securityGroup.id,
            fromPort: 3000,
            toPort: 3000,
            protocol: "TCP"
        }, this.options);
    }

    constructTaskArgs(): TaskDefinitionArgs {
        // TODO: we need to be aware of task vs container cpu/memory if more than 1 container will be present in task.
        //       Eg- using a sidecar pattern, fluent bit can ship logs to different locations, but now we have multiple consumers of resources. 
        const taskMemory = this.baseArgs.taskMemory ? this.baseArgs.taskMemory : 512;
        const taskCpu = this.baseArgs.taskCpu ? this.baseArgs.taskCpu : 256;
        
        const containerMemoryReservation = this.baseArgs.containerMemoryReservation ? this.baseArgs.containerMemoryReservation : 128;
        const containerCpu = this.baseArgs.containerCpu ? this.baseArgs.containerCpu : taskCpu;

        // set the desired state (count of tasks) for our ECS service
        const desiredNumberTasks = this.baseArgs.numberDesiredTasks ? this.baseArgs.numberDesiredTasks : 3;

        // from configuration, build the type of logger requested by the caller; awslogs, awsfirelens, splunk...
        let logDriver: LogDriver | undefined;
        if (this.baseArgs.logType != undefined) {

            logDriver = new LogFactory().buildLogDriver(
                this.baseArgs.logType,
                "console",
                this.baseArgs.logArgs,
                this.options);
        }

        // JSON will be definition provided to ECS for Console containers
        // Fully qualified ECR tag will be built for `image` property below
        const containerDefinitions = pulumi
            .all([
                this.baseArgs.accountId, 
                logDriver?.outputs, 
                this.baseArgs.trafficManager.console.loadBalancer.dnsName, 
                this.baseArgs.ecrRepoAccountId])
            .apply(([
                accountId, 
                logOutputs, 
                loadBalancerUrl, 
                ecrRepoAccountId]) => {

                const ecrAccountId = ecrRepoAccountId && ecrRepoAccountId !== "" ? ecrRepoAccountId : accountId;

                return JSON.stringify([{
                    name: consoleContainerName,
                    image: buildECRImageTag(ecrAccountId, this.baseArgs.region, this.consoleImage),
                    cpu: containerCpu,
                    memoryReservation: containerMemoryReservation,
                    portmappings: [{
                        containerPort: consolePort
                    }],
                    environment: this.constructEnvironmentVariables(loadBalancerUrl),
                    secret: [],
                    logConfiguration: logDriver && {
                        logDriver: LogType[logDriver!.logType],
                        // http://docs.aws.amazon.com/AmazonECS/latest/developerguide/using_awslogs.html#w2ab1c21c21c13
                        options: logDriver?.getConfiguration(logOutputs)
                    },
                }]);
            });

        // args will be given to the base container service to detail how Pulumi API (service) tasks should be constructed
        const taskArgs: TaskDefinitionArgs = {
            containerDefinitionArgs: containerDefinitions,
            numberDesiredTasks: desiredNumberTasks,
            cpu: taskCpu,
            memory: taskMemory,
            containerName: consoleContainerName,
            containerPort: consolePort
        };

        return taskArgs;
    }

    // Complete list of environment variables that each ECS task of Pulumi API (service) will inherit
    constructEnvironmentVariables(loadBalancerUrl: string): any[] {

        const {
            dns,
            region,
            recaptchaSiteKey,
            samlSsoEnabled,
            hideEmailSignup,
            hideEmailLogin
        } = this.baseArgs;

        return [
            {
                name: "PULUMI_API",
                value: `https://${dns.apiUrl}`
            },
            {
                name: "PULUMI_CONSOLE_DOMAIN",
                value: dns.consoleUrl,
            },
            {
                name: "PULUMI_HOMEPAGE_DOMAIN",
                value: this.baseArgs.dns.consoleUrl,
            },
            {
                name: "AWS_REGION",
                value: region
            },
            { 
                // enabling SSO requires that an org already be created and configured to use a SSO provider
                // https://github.com/pulumi/pulumi-service/pull/7953
                name: "SAML_SSO_ENABLED",
                value: samlSsoEnabled ? String(samlSsoEnabled) : ""
            },
            {
                name: "RECAPTCHA_SITE_KEY",
                value: recaptchaSiteKey
            },
            {
                // https://github.com/pulumi/pulumi-service/pull/7953
                name: "PULUMI_HIDE_EMAIL_LOGIN",
                value: hideEmailLogin ? String(hideEmailLogin) : ""
            },
            {
                // https://github.com/pulumi/pulumi-service/pull/7953
                name: "PULUMI_HIDE_EMAIL_SIGNUP",
                value: hideEmailSignup ? String(hideEmailSignup) : ""
            }
        ]
    }
}